R/teal_data_module-eval_code.R

```diff
@@ -0,0 +1,124 @@
+setOldClass("teal_data_module")
+
+#' Evaluate the code in the qenv environment
+#' @name eval_code
+#' @description
+#' Given code is evaluated in the `qenv` environment of `teal_data` reactive defined in `teal_data_module`.
+#' @param object (`teal_data_module`)
+#' @inheritParams teal.code::eval_code
+#' @return Returns a `teal_data_module` object.
+#' @importMethodsFrom teal.code eval_code
+#' @importFrom methods setMethod
+NULL
+
+#' @rdname eval_code
+#' @export
+#'
+#' @examples
+#' tdm <- teal_data_module(
+#'   ui = function(id) div(id = shiny::NS(id)("div_id")),
+#'   server = function(id) {
+#'     shiny::moduleServer(id, function(input, output, session) {
+#'       shiny::reactive(teal_data(IRIS = iris))
+#'     })
+#'   }
+#' )
+#' \dontrun{
+#' eval_code(tdm, "IRIS <- subset(IRIS, Species == 'virginica')")
+#' }
+setMethod("eval_code", signature = c("teal_data_module", "character"), function(object, code) {
+  teal_data_module(
+    ui = function(id) {
+      ns <- NS(id)
+      object$ui(ns("mutate_inner"))
+    },
+    server = function(id) {
+      moduleServer(id, function(input, output, session) {
+        teal_data_rv <- object$server("mutate_inner")
+
+        if (!is.reactive(teal_data_rv)) {
+          stop("The `teal_data_module` must return a reactive expression.", call. = FALSE)
+        }
+
+        eventReactive(teal_data_rv(),
+          {
+            data <- tryCatch(teal_data_rv(), error = function(e) e)
+
+            if (inherits(data, "teal_data")) {
+              eval_code(data, code)
+            } else if (inherits(data, "error")) {
+              data
+            } else {
```

Comment:
I don't understand when this clause is triggered.

Reply:
It is possible when `server` of `teal_data_module` returns error (not related to qenv.error). There is an example app which @averissimo provided:

https://github.com/insightsengineering/teal/pull/983#discussion_r1412098320


@chlebowa I think it might be removed tbh. When "evaluated" `teal_data_module` throws then `eventReactive` throws as well and error would be caught by `srv_teal_with_splash`.

Reply:
@averissimo I think we can remove `else if (inherits(data, "error")) {` and keep only `else`. Reason for this is that `inherits(data, "error")` expects this error to be already **handled**.  I expect handled errors to be very rare output of `teal_data_module` and we can throw them to the `else` or remove else at all and just return `data` (not evaluated).

Reply:
This clause catches both `shiny.silent.errors` and `qenv.errors` and pushes those up to be handled.

We could make it more specific to specifically catch those 2, however, we risk having error messages that look like this:

![image](https://github.com/insightsengineering/teal/assets/211358/06ade676-6864-4cf5-b8f5-2f8dfd93b58b)

Reply:
Problem with your app is that server returns objects of class errors but it doesn't really throw errors. If you 

There is a difference between
```t
structure(
  list(message = "🔴 is an object of class `error"),
  class = "error"
)

# and 
stop("this is error")
```

This app for example is done in a way which would never get inside of the `eventReactive` in `eval_code`. 

```r

tdm <- teal_data_module(
  ui = function(id) NULL,
  server = function(id) {
    moduleServer(id, function(input, output, session) {
      # Mock shiny.silent.error
      reactive(stop("error"))
    })
  }
)

tdm_new <- eval_code(tdm, "dataset1$id <- seq_len(nrow(dataset1))") |>
  within(
    dataset1$checksum <- apply(dataset1, 1, rlang::hash)
  )

init(
  data = tdm_new,
  modules = list(
    example_module()
  )
) |> shiny::runApp()
```

Same case with `shiny.error`

```r
tdm <- teal_data_module(
  ui = function(id) NULL,
  server = function(id) {
    moduleServer(id, function(input, output, session) {
      # Mock shiny.silent.error
      reactive(validate(need(FALSE, "this is shiny.error")))
    })
  }
)

tdm_new <- eval_code(tdm, "dataset1$id <- seq_len(nrow(dataset1))") |>
  within(
    dataset1$checksum <- apply(dataset1, 1, rlang::hash)
  )

init(
  data = tdm_new,
  modules = list(
    example_module()
  )
) |> shiny::runApp()

```

Reply:
It will probably be very rare as you said, but this way we allow for those ocassions to produce a readable message.